session-helper/flatpak-session-helper.c

```diff
@@ -425,24 +447,88 @@ file_changed (GFileMonitor     *monitor,
               GFile            *file,
               GFile            *other_file,
               GFileMonitorEvent event_type,
-              char             *source)
+              MonitorData      *data)
 {
-  if (event_type == G_FILE_MONITOR_EVENT_CHANGES_DONE_HINT ||
-      event_type == G_FILE_MONITOR_EVENT_CREATED)
-    copy_file (source, monitor_dir);
+  if (event_type != G_FILE_MONITOR_EVENT_CHANGES_DONE_HINT)
+    return;
+
+  char *real = realpath (data->source, NULL);
+  if (real)
+    {
+      /* detect the case that the file's real path has changed and
+         re-add the monitor for the real file */
+      if (g_strcmp0 (data->source, real) &&
+          g_strcmp0 (data->real, real))
+       {
+          free (data->real);
+          data->real = real;
+
+          if (data->monitor_real)
+            {
+              g_signal_handlers_disconnect_by_data (data->monitor_real, data);
+              g_clear_object (&(data->monitor_real));
+            }
+
+          GFile *r = g_file_new_for_path (real);
+          data->monitor_real = g_file_monitor_file (r, G_FILE_MONITOR_NONE, NULL, NULL);
+          if (data->monitor_real)
+            g_signal_connect (data->monitor_real, "changed", G_CALLBACK (file_changed), data);
+        }
+      else
+        {
+          /* detect the case that the source file is no longer a symlink and
+             remove the old real file monitor */
+          if (!g_strcmp0 (data->source, real))
+            {
+              free (data->real);
+              data->real = NULL;
+
+              if (data->monitor_real)
+                {
+                  g_signal_handlers_disconnect_by_data (data->monitor_real, data);
+                  g_clear_object (&(data->monitor_real));
+                }
+            }
+
+          free (real);
+        }
+    }
+
+  copy_file (data->source, monitor_dir);
 }
 
 static void
 setup_file_monitor (const char *source)
 {
   GFile *s = g_file_new_for_path (source);
-  GFileMonitor *monitor;
+  char *real;
+  GFileMonitor *monitor_source, *monitor_real = NULL;
+  MonitorData *data = NULL;
 
   copy_file (source, monitor_dir);
 
-  monitor = g_file_monitor_file (s, G_FILE_MONITOR_NONE, NULL, NULL);
-  if (monitor)
-    g_signal_connect (monitor, "changed", G_CALLBACK (file_changed), (char *) source);
+  monitor_source = g_file_monitor_file (s, G_FILE_MONITOR_NONE, NULL, NULL);
+  if (!monitor_source)
+    return;
+
+  real = realpath (source, NULL);
+  if (real && g_strcmp0 (source, real))
+    {
+      GFile *r = g_file_new_for_path (real);
+      monitor_real = g_file_monitor_file (r, G_FILE_MONITOR_NONE, NULL, NULL);
+    }
+
+  data = g_new0 (MonitorData, 1);
+  data->source = source;
```

Comment:
Can you guarantee that the lifetime of `source` is longer than the lifetime of the `MonitorData`? If not, this should be `g_strdup()`ed.

Reply:
It's a static const literal, so am pretty confident, yeah. :)

Reply:
Right, but there’s technically nothing in the declaration of `setup_file_monitor()` which requires the argument to be `static`. I think this could do with a comment at the top of `setup_file_monitor()` and maybe one on the `MonitorData` struct too, to prevent people misusing it in future.